lib/tree.rb

```diff
@@ -1,66 +1,135 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n) worst case
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    if current_node.nil?
+      return TreeNode.new(key, value)
+    end 
+    
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end 
+    return current_node
+  end 
+  
+  # Time Complexity: O(log(n) best case, O(n) worst case
+  # Space Complexity: O(1)
   def find(key)
```

Comment:
👍 , Since you have a recursive helper method you have a space complexity of O(log n) if the tree is balanced or O(n) if not.